radiosity.cpp

```diff
@@ -0,0 +1,327 @@
+                    if (conf.gamma_correction_enabled) {
+                        image(x, y).r = powf(image(x, y).r, conf.inverse_gamma);
+                        image(x, y).g = powf(image(x, y).g, conf.inverse_gamma);
+                        image(x, y).b = powf(image(x, y).b, conf.inverse_gamma);
+                    }
+                }
+            }));
+        }
+
+        long completed = 0;
+
+        for (auto& task: tasks) {
+            task.get();
+            completed += 1;
+            float progress = static_cast<float>(completed) / tasks.size();
+            int bar_width = progress * 20;
+            std::cerr
+                << "\rRendering          "
+                << "[" << std::string(bar_width, '-')
+                << std::string(20 - bar_width, ' ') << "] "
+                << std::setfill(' ') << std::setw(6)
+                << std::fixed << std::setprecision(2) << (progress * 100.0) << '%';
+            std::cerr.flush();
+        }
+        std::cerr << std::endl;
+
+        // Render feature lines after
+        // "Ray Tracing NPR-Style Feature Lines" by Choudhury and Parker.
+        std::cerr << "Drawing mesh lines ";
+        std::vector<std::future<void>> mesh_tasks;
+        float offset = 1.f;
+        std::vector<Vec2> offsets =
+            { Vec2(0.f, 0.f)
+            , Vec2(offset, 0.f)
+            , Vec2(offset, offset)
+            , Vec2(0.f, offset)
+            , Vec2(0.f, offset / 2)
+            , Vec2(offset / 2, offset)
+            , Vec2(offset, offset / 2)
+            , Vec2(offset / 2, 0.f)
+            };
+        for (int y = 0; y < height; ++y) {
+            mesh_tasks.emplace_back(pool.enqueue([
+                    &image, &offsets, &cam, &tree, &lights, width, height, y, &conf]()
+            {
+                for (int x = 0; x < width; ++x) {
+                    float dist_to_triangle, s, t;
+                    std::unordered_set<std::uintptr_t> triangle_ids;
+
+                    // Shoot center ray.
+                    auto cam_dir = cam.raster2cam(
+                        aiVector2D(x + 0.5f, y + 0.5f), width, height);
+                    auto center_id = reinterpret_cast<std::uintptr_t>(tree.intersect(
+                        Ray(cam.mPosition, cam_dir), dist_to_triangle, s, t));
+                    triangle_ids.insert(center_id);
+
+                    // Sample disc rays around center.
+                    // TODO: Sample disc with Poisson or similar.
+                    for ( auto offset : offsets) {
+                        cam_dir = cam.raster2cam(
+                            aiVector2D(x + offset[0], y + offset[1]), width, height);
+                        auto triangle_pt = reinterpret_cast<std::uintptr_t>(tree.intersect(
+                            Ray(cam.mPosition, cam_dir), dist_to_triangle, s, t));
+                        triangle_ids.insert(triangle_pt);
+                    }
+
+                    const float M_2 = 0.5f * offsets.size();
```

Comment:
Dann schlage ich vor std::vector durch constexpr std::array zu ersetzen. Anscheinend eigentlich sich aiVector2 dafür nicht, da es kein literal type ist. Hier sind die Anforderungen an solch einen Typ: http://stackoverflow.com/questions/21787488/can-have-definition-variable-of-non-literal-type-in-constexpr-function-body-c1. Es ist relativ einfach einen eigenen Typen zu schreiben, der genau das tut. Das einzige, was man in dem machen muss, ist den [] operator zu definieren. Schau dir die Deklaration von aiVector2 an.